### DISCUSSION TITLE
"Must be called" delegate parameters

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#4187] [Language Ideas]
Imagine this situation:
```
void Perform(Action action)
{
  PrepareStuff(); // this can be a few lines of code
  try
  {
    action();
    FinalizeOperation(); // this can be a few lines of code
  }
  catch(Exception ex)
  {
    RollbackPreparations(); // this can be a few lines of code
  }
}
...
string s, s2 = null!;
Perform(()=> { s="abc"; s2="def"; });
Console.Write(s); // error - s was not assigned
Console.Write(s2); // OK - workaround, assigning null, because WE KNOW that Perform will call action delegate
```
This pattern is helpful to reuse code.
Can we add something like below, using a new **mustcall** keyword:
```
void Perform(mustcall Action action)
{
  PrepareStuff(); // this can be a few lines of code
  try
  {
    action();
    FinalizeOperation(); // this can be a few lines of code
  }
  catch(Exception ex)
  {
    RollbackPreparations(); // this can be a few lines of code
  }
}
...
string s;
Perform(()=> { s="abc"; });
Console.Write(s); // OK - s was assigned in delegate, C# knows delegate is called because of mustcall
...
void Foo(mustcall Action action) // error: action is not called in "non-conditional execution flow"
{
  if(...)
  {
    action();
  }
}
```

**Comment by HaloFour** [comment]
This sounds like this could be accomplished with a custom attribute and an analyzer which would error if the delegate was not called on every code path.

**Reply by TahirAhmadov** [reply]
Can an analyzer "cancel" the built in error of "local variable s was not assigned"?

**Reply by HaloFour** [reply]
No, an analyzer couldn't do that.

There are numerous other ways that this method could assign those variables that would work with definite assignment.  Is there a reason that these are not suitable here?

**Reply by TahirAhmadov** [reply]
I'm unclear what methods you are referring to. I could abolish `Perform` and instead package its stages into multiple other methods, then inline the entire thing, such as:

```
void Perform_Prepare() { ...}
void Perform_Finalize() { ... }
void Perform_Error(Exception ex) { ... }
.....
string s;
Perform_Prepare();
try
{
  s = "abc"; // this used to be the body of action
  Perform_Finalize();
}
catch(Exception ex)
{
  Perform_Error(ex);
}
```

But you can see how this way more cumbersome and inelegant.

**Reply by HaloFour** [reply]
You can have `Perform` accept a delegate that returns a value and then have `Perform` return that value.

```cs
T Perform<T>(Func<T> action)
{
  PrepareStuff(); // this can be a few lines of code
  try
  {
    T value = action();
    FinalizeOperation(); // this can be a few lines of code
    return value;
  }
  catch(Exception ex)
  {
    RollbackPreparations(); // this can be a few lines of code
  }
}
...
var (s, s2) = Perform(() => { return ("abc", "def"); });
Console.Write(s); // OK
Console.Write(s2); // OK
```

**Reply by TahirAhmadov** [reply]
Yes, this can work, but again, it's not elegant: I now have to exclusively use `Func<T>` instead of `Action` for all delegate parameters used as "main execution flow callbacks", and if I already have a `Func<T>`, I need to wrap it like so: `Func<Tuple<TReturn, TDefAssign>>` 

I actually prefer my workaround to yours, because it requires fewer deviations from syntax which describes the intent of the code.

Also, the **mustcall** isn't just about variable assignment. It allows you to specify that C# compiler "inserts" the execution flow of the delegate into the flow of the method call, similarly to local functions; the feature parity here creates a cohesive architecture, IMO. I can't think of any other benefits of adding this feature, that I personally would make use of at the moment, but I'm sure others will.
PS. I guess one obvious improvement would be to nullability determination.

```
void foo()
{
  string s;
  void bar()
  {
    s = "abc";
  }
  if(...)
    bar();
  else
    Perform(bar);
  Console.Write(s); // OK - local function already works, and the new mustcall will allow the same effect for Perform call
}
```

**Reply by HaloFour** [reply]
What I suggested is not a workaround, it's the idiomatic way to approach a problem like this and I would argue is much more elegant.  My approach flows the values back from the method to the variable assignment vs. relying on the assignment as a side effect.  It also has the advantage of already working vs. requiring changes to the language, which is massively expensive.

**Reply by TahirAhmadov** [reply]
What you wrote would be true if both `Perform` and the code calling it were revolving around returning value(s). However in my example of a procedural invocation, there are no values being returned; `Perform` has no idea if the caller sets any variables as part of its executions - some callers may need to do it, others may not. The "intent" is for it to be `void`, not `T`. The fact that assignment is a side effect is by design.

```
Func<T> Perform<T>(Func<T> action) { .... return action(); ... } // this is your suggestion
......
Perform<bool>(()=>
{
  ...... // do stuff; no need to assign any variables for our scenario
  return false; // return some dummy value
});
.....
Action Perform(Action action) { .... action(); ... } // or we have to duplicate such methods

```

**Reply by HaloFour** [reply]
You could have an overload to `Perform` which accepts an `Action` for those cases where the consumer does not need a return value.

> The fact that assignment is a side effect is by design.

So is the fact that assignment by delegate executed by another method does not definitely assign a variable.  I'd suggest that if this is the design that you prefer that you already have the workaround of assigning those variables at declaration with a sentinel value.

**Reply by TahirAhmadov** [reply]
> 
> So is the fact that assignment by delegate executed by another method does not definitely assign a variable. 

I know it is - currently - which is why I suggested a change. 

I recall reading that one of the benefits of using local functions was the ability of the compiler to correctly traverse the execution flow (for all of its intents - including definitive assignment). All I'm saying is that we can accomplish the same for the often used callback situation. If it was deemed a benefit - that local functions allow the compiler to traverse the execution flow - wouldn't the same reasoning apply here? If it's a benefit for local functions getting called locally, it's also a benefit for delegates being called remotely (given the **mustcall** keyword in the method signature)?

**Reply by HaloFour** [reply]
A local function is a part of the method in which it is defined and thus can participate in that execution flow.  Once you call out of the current method to another method you are no longer in that flow.

I get that you're asking for this to be changed which is why you opened this discussion, but given that there are several other (and more idiomatic) approaches to solving this kind of problem using existing syntax I'd suggest that it's very unlikely to be considered.

**Reply by TahirAhmadov** [reply]
> A local function is a part of the method in which it is defined and thus can participate in that execution flow. 

Before being introduced, we lived happily without local functions: local delegates solved most of the same tasks, except for a few small differences, obviously. Apparently it was deemed important enough to have that execution flow traversal (in addition to other benefits) to go through the expense of adding local functions.
Slightly off topic, to be honest, I think in terms of implementation cost vs benefit to programmers, local functions give you less bang for your buck, than my suggestion. Having a method with nested local functions is on a path to having bloated methods. We want to encourage people to outsource logic to other reusable methods, so that each method is small and provides incremental piece of the code puzzle. Local functions help the "large method with many nested local functions" anti-pattern and do nothing for the "modularize and reuse code by splitting it into many small methods" DRY, IMHO much more preferable, approach. 
And even WRT potential performance benefits of local functions, I would prefer method inlining at optimization stage; the code is clean and modularized, and then we make it fast by inlining strategically; win-win.

> Once you call out of the current method to another method you are no longer in that flow.

Yes - unless you use something like **mustcall** to "tell" C# that you are, in fact, in that flow.

**Reply by TahirAhmadov** [reply]
On the same topic, I'm not sure if it happens already, but it would be cool if C# was able to inline both the hypothetical `Perform` and the `action` (given it fits the usual criteria, like size, etc.) - regardless of `mustcall`, at compile time. Then instead of creating a closure state machine, a separate method, all the overhead of passing around delegates and indirect invocation, things like `IEnumerable<>.First(x=>...)` become much faster: they get output as a simple `foreach` with the condition.

**Reply by HaloFour** [reply]
> Yes - unless you use something like **mustcall** to "tell" C# that you are, in fact, in that flow.

It's not in that flow.  The `Perform` method is a total blackbox.  The caller has no idea what it does, nor should it.  Your flag doesn't change this, it only tells the caller to assume that the delegate should have been called at some point and to make assumptions about what side effects may have resulted.

> but it would be cool if C# was able to inline

The C# compiler does not perform such optimizations.  It produces IL that captures the behavior of the code as it was written and expects that the runtime will handle optimizations such as inlining.  The version of the assemblies used at compile time could be different from those used at runtime.  The compile time assemblies could be reference assemblies which lack method bodies.

**Reply by TahirAhmadov** [reply]
> 
> 
> > Yes - unless you use something like **mustcall** to "tell" C# that you are, in fact, in that flow.
> 
> It's not in that flow. The `Perform` method is a total blackbox. The caller has no idea what it does, nor should it. Your flag doesn't change this, it only tells the caller to assume that the delegate should have been called at some point and to make assumptions about what side effects may have resulted.

We rely on flags like this already; prime example - `out` as opposed to `ref`. It's a flag which says the called method must assign a value to the parameter, and the calling site relies on that flag to expect that a value is assigned. I can reword your paragraph and replace "delegate" with "parameter", and "called" with "assigned", and the same logic would apply to `out` parameters - which I think you'd agree is invalid. So if we have `out` flag in addition to `ref`, it's valid to have `mustcall`: in both cases, it's a signature which enforces something inside the method and promises something to the caller.

> 
> > but it would be cool if C# was able to inline
> 
> The C# compiler does not perform such optimizations. It produces IL that captures the behavior of the code as it was written and expects that the runtime will handle optimizations such as inlining. The version of the assemblies used at compile time could be different from those used at runtime. The compile time assemblies could be reference assemblies which lack method bodies.

True, I thought about that right after posting my comment; compile time wouldn't work. I wrote it like that because I was worried about the complexity of this optimization at run time; the optimizer would need to reverse engineer the whole "closure state machine" construct, and revert it to regular method with stack-stored locals; to do that, it would need to make sure those state machine "locals" are used only in the body of the calling method and in the optimized lambdas, but not any non-optimized lambdas, before doing the optimization; very complicated. On the plus side, even if the locals are used in other lambdas, and state machine stays, it's still faster - we get rid of `First` calling our predicate lambda in a loop. Anyway, this is entirely off-topic.

**Reply by HaloFour** [reply]
`out` parameters being another way you could accomplish this without language changes.

Yes, you can make the argument that this proposal is similar to `out` parameters, but it's also substantially less useful, especially since it explicitly encourages relying on side effects from lambdas which is an anti-pattern.

**Reply by TahirAhmadov** [reply]
> 
> 
> `out` parameters being another way you could accomplish this without language changes.

Yes - if `Perform` was about returning values, like `int.TryParse`; it is not.

> 
> Yes, you can make the argument that this proposal is similar to `out` parameters, but it's also substantially less useful, especially since it explicitly encourages relying on side effects from lambdas which is an anti-pattern.

The word "side effect" I think is not a good description of what's happening here. 
Correct usage of the term: modifying/invoking a non-public member using reflection can have side effects - an unexpected effect outside of your intended behavior. 
In my example, a) `Perform` does its thing: preparation, callback, finalization, reverting on error; it has no side effects; and b) the calling site assigns a closure local variable on purpose - it's not a side effect; it's intended behavior.

**Reply by HaloFour** [reply]
@TahirAhmadov 

You are asking `Perform` to modify state outside of its local environment.  That is _the_ definition of side effect.

**Reply by TahirAhmadov** [reply]
> 
> 
> @TahirAhmadov
> 
> You are asking `Perform` to modify state outside of its local environment. That is _the_ definition of side effect.

Right, the calling site is _asking_ for it. How can something that is explicitly asked for, be a side effect?

Any IoC approach is susceptible to it; if we didn't want to have methods like `Perform`, then why have delegates/function pointers altogether? All those can modify the state outside of `Perform`, regardless of `mustcall`.

**Reply by HaloFour** [reply]
> How can something that is explicitly asked for, be a side effect?

It's not the intent of `Perform` to cause an assignment or a state change.  That's what makes it a side effect.  The fact that the consumer of `Perform` explicitly intends this doesn't change it.  You can do the same thing with LINQ and it's considered very much an anti-pattern.  If you intend the API to flow state the API should be designed to flow state.  There are multiple ways to accomplish this without requiring any language changes.

**Reply by HaloFour** [reply]
I feel like we're going in circles and not going to come to an agreement.  Anyway, if you want to discuss this informally with other members of the repo you can join the Gitter channel here:  https://gitter.im/dotnet/csharplang 😄

**Reply by TahirAhmadov** [reply]
> 
> 
> > How can something that is explicitly asked for, be a side effect?
> 
> It's not the intent of `Perform` to cause an assignment or a state change. That's what makes it a side effect. The fact that the consumer of `Perform` explicitly intends this doesn't change it. You can do the same thing with LINQ and it's considered very much an anti-pattern. If you intend the API to flow state the API should be designed to flow state. There are multiple ways to accomplish this without requiring any language changes.

Something being an anti-pattern is not the same as something having a side effect. "Anti pattern" means a pattern of code which leads to undesirable outcomes of some sort. I'd agree that LINQ lambdas modifying the state is an anti-pattern; they're intended to be simple functions, like predicates, key selectors, etc. Turning these into full fledged delegates which modify the state doesn't bode well for reliability of the code, for various reasons. But even those don't have "side effects": the calling site intends to modify some closure local for whatever reason, and that's the only effect it gets; that's not a side effect.
`Perform` expects an `Action`, which has "procedural" style; procedures do modify the state; that's one of the key differences between functions and procedures. Therefore, it's not an anti pattern for "procedural" call backs to modify the state.

**Reply by HaloFour** [reply]
No, specifically the reliance of side effects in this case would be the anti-pattern and it's not something that the language should be modified to encourage.

We're simply not going to agree on this point.  The fact that there are at least three different ways to accomplish this today without any language changes makes it not worthwhile to me.  They might not be as ergonomic as your preference, but I don't think that this is a case that is remotely common enough or should be encouraged to be more ergonomic.

IMO, the most elegant solution here us to use a `Func<T>` and allow `Perform` to appropriately flow state back to the caller.  For those cases where you don't need/want to flow state you could have an overload that accepts an `Action`.

**Reply by theunrepentantgeek** [reply]
> The word "side effect" I think is not a good description of what's happening here.

It’s **_exactly_** the right description. 

“Side effect” is the correct term for any function (or delegate) that makes a change to external state. 

It’s not about whether the side effect is desirable or intended. 

You’re writing code that relies on side effects to work correctly - introducing both temporal and indirect coupling to your code. 

Were I to see code like this in a code review, I’d push back and request the design be reconsidered, as the longer term maintenance headaches cause by this style of coding are significant.

**Reply by TahirAhmadov** [reply]
> 
> 
> > The word "side effect" I think is not a good description of what's happening here.
> 
> It’s **_exactly_** the right description.
> 
> “Side effect” is the correct term for any function (or delegate) that makes a change to external state.
> 
> It’s not about whether the side effect is desirable or intended.
> 
> You’re writing code that relies on side effects to work correctly - introducing both temporal and indirect coupling to your code.
> 
> Were I to see code like this in a code review, I’d push back and request the design be reconsidered, as the longer term maintenance headaches cause by this style of coding are significant.

So which function or delegate changes external state in my scenario? 
`Perform` doesn't - it just calls a supplied delegate. 
The `action` delegate doesn't - the closure local is not external to it, it's in the same scope.

**Reply by HaloFour** [reply]
> So which function or delegate changes external state in my scenario?

The state is local to the method in which the `action` delegate is declared, not to the `action` delegate itself.  If it was local state to the delegate, those variables would be declared within the delegate body and you would not require/rely on closure state, nor would the mutation be observable after either `Perform` or `action` was invoked.  This is a textbook example of side effects.

**Reply by TahirAhmadov** [reply]
> 
> 
> > So which function or delegate changes external state in my scenario?
> 
> The state is local to the method in which the `action` delegate is declared, not to the `action` delegate itself. If it was local state to the delegate, those variables would be declared within the delegate body and you would not require/rely on closure state, nor would the mutation be observable after either `Perform` or `action` was invoked. This is a textbook example of side effects.

That is the most bizarre statement I've ever heard. 
If closure state is external to the delegate, and thus shouldn't be accessed, then why is closure a thing at all (in any language, not just C#)?
What about local functions, why are they allowed to access the locals defined in the enclosing method?
What about class fields, are they external, too?
What about modifying method scope locals inside an if/while/etc. statement block?
Let's go even further, isn't throwing exception having a side effect, as it bubbles up the call chain and wrecks everything in its path?

**Reply by HaloFour** [reply]
> If closure state is external to the delegate, and thus shouldn't be accessed

Accessing it is fine.  Mutating it is where it becomes questionable.  Requiring it to be mutated as part of the execution flow is, well, inadvisable.

> What about class fields, are they external, too?

To the delegate?  Yes.

> Let's go even further, isn't throwing exception having a side effect, as it bubbles up the call chain

No, exceptions aren't really any different than a return value.  They are state passed explicitly from the callee to the caller and that state is not observable unless it continues to bubble up the chain.

**Reply by TahirAhmadov** [reply]
> > If closure state is external to the delegate, and thus shouldn't be accessed
> 
> Accessing it is fine. Mutating it is where it becomes questionable. Requiring it to be mutated as part of the execution flow is, well, inadvisable.

If the problem is modifying a variable, then it's an argument for a purely functional language, which we all agree, C# isn't.
If the problem is accessing closure locals at all, then it conflicts with every language under the sun which has closures.
If modifying variables is fine, and closure locals are fine, why is modifying closure locals not good? 
Furthermore, local functions do the "execution flow insertion" to analyze assignments/nullability/etc. So the language is perfectly fine with local functions modifying method locals; local functions were introduced relatively recently, so it cannot be chalked up to "legacy". If modifying method locals was a no-no, why weren't local functions barred from doing so?
Again, I suspect you are very strongly opinionated against modifying closure locals from inside LINQ lambdas, which I agree with - because they're executed in a loop, go against the LINQ style, etc. But my scenario just isn't it.

> > What about class fields, are they external, too?
> 
> To the delegate? Yes.

No, I meant to a class method (forget delegates for a moment). 
The state is local to the _class_ in which the _method_ is declared, not to the _method_ itself. If it was local state to the _method_, those _fields_ would be declared within the _method_ body and you would not require/rely on _this keyword_, nor would the mutation be observable after either _method_ was invoked. This is a textbook example of side effects. 
See what I did there?
> 
> > Let's go even further, isn't throwing exception having a side effect, as it bubbles up the call chain
> 
> No, exceptions aren't really any different than a return value. They are state passed explicitly from the callee to the caller and that state is not observable unless it continues to bubble up the chain.

Exceptions are very different from a return value; imagine that A called B, and B called C, and C throws, now A has to deal with an exception from a method it has no idea about - which is a side effect. (Just to be clear, I'm making a rhetorical argument to try communicating to you how wrong I think your position is on this.)

**Comment by HaloFour** [comment]
> If the problem is modifying a variable, then it's an argument for a purely functional language, which we all agree, C# isn't.

The problem is the way through which the variable is modified.

> The state is local to the class in which the method is declared, not to the method itself. If it was local state to the method, those fields would be declared within the method body and you would not require/rely on this keyword, nor would the mutation be observable after either method was invoked. This is a textbook example of side effects.

Yes it is.  But that is the purpose of mutator methods.  However, you wouldn't expect a completely different method external to that class to result in mutation to that class.

> Exceptions are very different from a return value; imagine that A called B, and B called C, and C throws, now A has to deal with an exception from a method it has no idea about - which is a side effect.

Yes, exceptions can be considered a side effect if you observe it and change flow control.  The goal isn't to remove side effects, especially in C# which is, at best, a hybrid functional language.  It's to mitigate reliance on them and the "spooky action at a distance" which can easily cause programs to become difficult to reason about and maintain.

Anyway, you can accomplish everything you want to do, even in the way you want to do it.  You want it to be slightly easier via changing the language so that you can short-circuit definite assignment.  Any language change needs to be justified due to the exorbitant costs via demonstrating wide benefit.  It's my opinion that this proposal doesn't come close to that as it only saves a handful of keystrokes in an incredibly niche case that can be implemented through more idiomatic means, like a return value.

**Reply by TahirAhmadov** [reply]
> Yes it is. But that is the purpose of mutator methods. However, you wouldn't expect a completely different method external to that class to result in mutation to that class.

Right, and a delegate just does not fit the definition of a "completely different method external to that method".
It may be possible with unsafe context, I haven't tried it, but the analogy is, a) using reflection to modify private fields of another class, and b) using unsafe code to find the location of a local in a method frame up the stack and modify its value by writing to memory with the pointer. 
Both of these are obviously terrible hacks which can have side effects. Notice how b) is starkly different from mutating a closure local.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> Right, and a delegate just does not fit the definition of a "completely different method external to that method".

Passing a lambda for the purpose of mutating enclosed state does fit that definition, in my opinion.

**Reply by TahirAhmadov** [reply]
> Passing a lambda for the purpose of mutating enclosed state does fit that definition, in my opinion.

You are obviously entitled to your opinion, but with all due respect, you haven't been able to substantiate it in any way. Having said that, let's just agree to disagree on this. I wrote everything that there was to write about this.

**Reply by HaloFour** [reply]
Not liking my opinion doesn't make it unsubstantiated.  As stated, what you're doing is the [literal definition](https://en.wikipedia.org/wiki/Side_effect_(computer_science)#:~:text=In%20computer%20science%2C%20an%20operation,the%20invoker%20of%20the%20operation.) of what a side effect is.

**Reply by TahirAhmadov** [reply]
> 
> 
> Not liking my opinion doesn't make it unsubstantiated. As stated, what you're doing is the [literal definition](https://en.wikipedia.org/wiki/Side_effect_(computer_science)#:~:text=In%20computer%20science%2C%20an%20operation,the%20invoker%20of%20the%20operation.) of what a side effect is.

That article is just wrong. Like I wrote previously, by that definition, the entire C# is based around side effects, and if, as you wrote, "it's inadvisable", we should abolish the entire C# language and transition to functional programming.

For example, the article lists "modifying a mutable argument passed by reference" as an example of a side effect. Passing in arguments by reference is intentional, by the function signature, by the caller, etc. If exceptions are like a return value, so are by-ref args.

This is easily one of the worst Wikipedia articles (outside of the heavily biased articles dealing with political matters). Please, please don't rely on Wikipedia. Listen to me instead :)

**Reply by HaloFour** [reply]
> That article is just wrong.

That would be your opinion, and it is certainly unsubstantiated.

> Like I wrote previously, by that definition, the entire C# is based around side effects

A decent amount of it is, particularly the OOP aspects.  That doesn't mean that reliance of side effects in the functional aspects of the language should be encouraged any more than they already are.  Changing the language specifically to support it better for your use case is a statement of endorsement by the language design team.

> This is easily one of the worst Wikipedia articles 

Pick a source.  Any source.  They all say literally the same thing because side effects in computer science are very well defined.

**Reply by TahirAhmadov** [reply]
> Pick a source. Any source. They all say literally the same thing because side effects in computer science are very well defined.

That's not how I approach challenges. I do listen and read to many sources, but at the end of the day, I form my own opinion.

I think what they mean by "side effect" is any "effect", any "mutation". The qualifier "side" means that there is an intended effect, and that there are other, "side" effect(s). The qualifier "side" by definition introduces the concept of intent into the equation.

Do this logical exercise: imagine a piece of code, which does not have any of the "side effects", as listed in the Wiki article, and only has "**non**-side" effects; and try to come up with examples of what those effects would include.

**Reply by HaloFour** [reply]
> That's not how I approach challenges. I do listen and read to many sources, but at the end of the day, I form my own opinion.

It becomes impossible to have a conversation on any subject if you're going to redefine the meaning of words to suit your preferences.

> The qualifier "side" by definition introduces the concept of intent into the equation.

Incorrect, effects are "side effects" irrespective of whether or not they were intended.  They may be the primary intention.  They are a "side effect" specifically virtue of observably mutating anything other than the primary inputs or outputs of the method.

> Do this logical exercise: imagine a piece of code, which does not have any of the "side effects", as listed in the Wiki article, and only has "non-side" effects; and try to come up with examples of what those effects would include.

That describes quite literally every "pure" function ever written.  It's a common practice, even in C#.  Relegate the side effects to the edges.  Of course in those cases the "side effects" are usually input and output from the program itself so they can't be eliminated.  In your case it is painfully trivial to do so.

If you want to argue over the tenets of functional programming I'd say we should take the conversation to [Gitter](https://gitter.im/dotnet/csharplang).

**Comment by CyrusNajmabadi** [comment]
@TahirAhmadov i'd recommend maybe bringing this to a better venue to dive deep into this idea.  Both gitter and discord have good forums for this.  Check out the links for them here: https://github.com/dotnet/roslyn#questions

**Comment by TahirAhmadov** [comment]
Update on this: I was able to find a better solution for one of my main particular use cases, specifically - ability to process exceptions. I used to have a method which calls a delegate inside a try-catch and the catch does something different than what happens after the delegate succeeds. Now, I use https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception?view=net-6.0 to achieve the same with `using` and a special `IDisposable`.
PS. `using` and `IDisposable` have the added benefit of reducing the number of stackframes, which helps with logging, debugging, etc.
